simple_parsing/helpers/serialization/serializable.py

```diff
@@ -792,6 +846,10 @@ def from_dict(
     for name, value in non_init_args.items():
         logger.debug(f"Setting non-init field '{name}' on the instance.")
         setattr(instance, name, value)
+
+    # Call __post_init__ method of the instance
```

Comment:
Hmmm I don't think this is necessary, since the generated `__init__` or `__new__` from `dataclasses.dataclass` already calls `__post_init__`.

Reply:
This is necessary because the `__post_init__` is after all init.
```python
    try:
        instance = cls(**init_args)  # type: ignore
    except TypeError as e:
        # raise RuntimeError(f"Couldn't instantiate class {cls} using init args {init_args}.")
        raise RuntimeError(
            f"Couldn't instantiate class {cls} using init args {init_args.keys()}: {e}"
        )

    for name, value in non_init_args.items():
        logger.debug(f"Setting non-init field '{name}' on the instance.")
        setattr(instance, name, value)
```
This snippet call `__init__` then, update with the `non_init_args`. This might not agree with the result of `__post_init__`. If I get rid of the calling of `__post_init__`, the test  **test/test_replace.py::test_replace_nested_dict[simple-config2-changes2]** would fail.

Reply:
I am little confused now. I checked the `dataclass.replace`. It raise `ValueError` when `init=False`. 
```python
if not f.init:
    # Error if this field is specified in changes.
    if f.name in changes:
        raise ValueError(f'field {f.name} is declared with '
                         'init=False, it cannot be specified with '
                         'replace()')
    continue
```

That contradicts with `from_dict` which allow assign value to field whose `init=False`.

Reply:
Wait, so when we're calling `cls(**init_args)` this **doesn't** call `cls.__post_init__()`? I'm very confused as well!

I think we should probably not set the `init=False` fields in `from_dict` also. But let's remove that in another PR, if that's simpler.

Reply:
I can confirm on my end, with the code of `from_dict` that is currently on the `master` branch:
`cls(**init_args)` **does** call __post_init__:
```python

def test_from_dict_calls_post_init():
    @dataclass
    class Bob:
        a: int = 123
        # Note: Using a default of None so we that if post_init isn't called, we can easily tell.
        negative_a: Optional[int] = field(init=False, default=None)

        def __post_init__(self):
            print(f"Called post-init!")
            self.negative_a = -self.a

    a = 1
    bob = Bob(a=a)
    assert bob.negative_a == -a
    assert to_dict(bob) == {
        "a": a,
        "negative_a": -a,
    }

    constructed_from_dict = from_dict(Bob, {"a": 111})
    assert constructed_from_dict == Bob(a=111)
    assert constructed_from_dict.negative_a == -111
```

As for the non-init fields, leave it as-is, and I'll make another PR to remove this feature after.